File: test/py-stub-outcome-overlay/BaboonTests/WiringTests.py
import unittest
import json
from io import BytesIO

from BaboonDefinitions.Generated.baboon_service_wiring import (
    BaboonMethodId, NoMatchingMethod, CallFailed, DecoderFailed,
)
from BaboonDefinitions.Generated.baboon_codecs import BaboonCodecContext
from BaboonDefinitions.Generated.baboon_runtime_shared import LEDataInputStream, LEDataOutputStream
from BaboonDefinitions.Generated.testpkg.pkg0.I1 import I1
from BaboonDefinitions.Generated.testpkg.pkg0.I2 import I2
from BaboonDefinitions.Generated.testpkg.pkg0.I1_Wiring import invoke_json_I1, invoke_ueba_I1
from BaboonDefinitions.Generated.testpkg.pkg0.I2_Wiring import invoke_json_I2, invoke_ueba_I2
from BaboonDefinitions.Generated.testpkg.pkg0.i1.testcall.In import In as I1_In, In_UEBACodec as I1_In_UEBACodec
from BaboonDefinitions.Generated.testpkg.pkg0.i1.testcall.Out import Out as I1_Out, Out_UEBACodec as I1_Out_UEBACodec
from BaboonDefinitions.Generated.testpkg.pkg0.i1.testcall.Err import Err as I1_Err
from BaboonDefinitions.Generated.testpkg.pkg0.T7_Empty import T7_Empty, T7_Empty_UEBACodec
from BaboonDefinitions.Generated.testpkg.pkg0.i2.noerrcall.In import In as I2_In, In_UEBACodec as I2_In_UEBACodec
from BaboonDefinitions.Generated.testpkg.pkg0.i2.noerrcall.Out import Out as I2_Out, Out_UEBACodec as I2_Out_UEBACodec
from BaboonTests.custom_containers import OutcomeSuccess, OutcomeFailure, OutcomeServiceRt


class MockI1(I1):
    def testCall(self, arg):
        return OutcomeSuccess(I1_Out(i00=42))

    def testCall2(self, arg):
        return OutcomeSuccess(T7_Empty())


class FailingI1(I1):
    def testCall(self, arg):
        return OutcomeFailure(I1_Err(msg="domain error"))

    def testCall2(self, arg):
        return OutcomeFailure(T7_Empty())


class ThrowingI1(I1):
    def testCall(self, arg):
        raise RuntimeError("service error")

    def testCall2(self, arg):
        raise RuntimeError("service error")


class MockI2(I2):
    def noErrCall(self, arg):
        return I2_Out(result="result_" + str(arg.value))


class OutcomeWiringTests(unittest.TestCase):

    ctx = BaboonCodecContext.default()
    rt = OutcomeServiceRt()

    # ==================== I1 JSON Tests ====================

    def test_i1_json_testCall_success(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "testCall")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeSuccess)
        decoded = json.loads(result.value)
        self.assertEqual(decoded["i00"], 42)

    def test_i1_json_testCall2_success(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "testCall2")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeSuccess)

    def test_i1_json_domain_error(self):
        impl = FailingI1()
        method = BaboonMethodId("I1", "testCall")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeFailure)
        self.assertIsInstance(result.error, CallFailed)

    def test_i1_json_no_matching_method(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "nonexistent")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeFailure)
        self.assertIsInstance(result.error, NoMatchingMethod)

    def test_i1_json_decoder_failed(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "testCall")
        result = invoke_json_I1(method, "not valid json!!", impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeFailure)
        self.assertIsInstance(result.error, DecoderFailed)

    def test_i1_json_service_throws(self):
        impl = ThrowingI1()
        method = BaboonMethodId("I1", "testCall")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeFailure)
        self.assertIsInstance(result.error, CallFailed)

    # ==================== I1 UEBA Tests ====================

    def test_i1_ueba_testCall_success(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "testCall")
        buf = BytesIO()
        writer = LEDataOutputStream(buf)
        I1_In_UEBACodec.instance().encode(self.ctx, writer, I1_In())
        input_bytes = buf.getvalue()
        result = invoke_ueba_I1(method, input_bytes, impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeSuccess)
        reader = LEDataInputStream(BytesIO(result.value))
        decoded = I1_Out_UEBACodec.instance().decode(self.ctx, reader)
        self.assertEqual(decoded.i00, 42)

    def test_i1_ueba_no_matching_method(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "nonexistent")
        result = invoke_ueba_I1(method, b'', impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeFailure)
        self.assertIsInstance(result.error, NoMatchingMethod)

    def test_i1_ueba_service_throws(self):
        impl = ThrowingI1()
        method = BaboonMethodId("I1", "testCall")
        buf = BytesIO()
        writer = LEDataOutputStream(buf)
        I1_In_UEBACodec.instance().encode(self.ctx, writer, I1_In())
        input_bytes = buf.getvalue()
        result = invoke_ueba_I1(method, input_bytes, impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeFailure)
        self.assertIsInstance(result.error, CallFailed)

    # ==================== I2 JSON Tests ====================

    def test_i2_json_success(self):
        impl = MockI2()
        method = BaboonMethodId("I2", "noErrCall")
        result = invoke_json_I2(method, '{"value": 123}', impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeSuccess)
        decoded = json.loads(result.value)
        self.assertEqual(decoded["result"], "result_123")

    # ==================== I2 UEBA Tests ====================

    def test_i2_ueba_success(self):
        impl = MockI2()
        method = BaboonMethodId("I2", "noErrCall")
        buf = BytesIO()
        writer = LEDataOutputStream(buf)
        I2_In_UEBACodec.instance().encode(self.ctx, writer, I2_In(value=456))
        input_bytes = buf.getvalue()
        result = invoke_ueba_I2(method, input_bytes, impl, self.rt, self.ctx)
        self.assertIsInstance(result, OutcomeSuccess)
        reader = LEDataInputStream(BytesIO(result.value))
        decoded = I2_Out_UEBACodec.instance().decode(self.ctx, reader)
        self.assertEqual(decoded.result, "result_456")


if __name__ == "__main__":
    unittest.main()

File: test/rs-stub-outcome-overlay/tests/wiring_tests.rs
use baboon_rs_stub::baboon_runtime::{BaboonCodecContext, BaboonBinEncode, BaboonBinDecode};
use baboon_rs_stub::baboon_service_wiring::{BaboonMethodId, BaboonWiringError};
use baboon_rs_stub::custom_containers::{Outcome, OutcomeServiceRt};
use baboon_rs_stub::testpkg::pkg0::i1::I1;
use baboon_rs_stub::testpkg::pkg0::i1::testcall::r#in::In as I1_testCall_in;
use baboon_rs_stub::testpkg::pkg0::i1::testcall::out::Out as I1_testCall_out;
use baboon_rs_stub::testpkg::pkg0::i2::I2;
use baboon_rs_stub::testpkg::pkg0::i2::noerrcall::r#in::In as I2_noErrCall_in;
use baboon_rs_stub::testpkg::pkg0::i2::noerrcall::out::Out as I2_noErrCall_out;
use baboon_rs_stub::testpkg::pkg0::t7_empty::T7_Empty;
use baboon_rs_stub::testpkg::pkg0::{invoke_json_I1, invoke_ueba_I1, invoke_json_I2, invoke_ueba_I2};

struct MockI1;
impl I1 for MockI1 {
    fn test_call(&self, _arg: I1_testCall_in) -> Outcome<I1_testCall_out> {
        Outcome::Success(I1_testCall_out { i00: 42 })
    }
    fn test_call2(&self, _arg: T7_Empty) -> Outcome<T7_Empty> {
        Outcome::Success(T7_Empty {})
    }
}

struct FailingI1;
impl I1 for FailingI1 {
    fn test_call(&self, _arg: I1_testCall_in) -> Outcome<I1_testCall_out> {
        Outcome::Failure(Box::new(baboon_rs_stub::testpkg::pkg0::i1::testcall::err::Err { msg: "domain error".to_string() }))
    }
    fn test_call2(&self, _arg: T7_Empty) -> Outcome<T7_Empty> {
        Outcome::Failure(Box::new(T7_Empty {}))
    }
}

struct MockI2;
impl I2 for MockI2 {
    fn no_err_call(&self, arg: I2_noErrCall_in) -> I2_noErrCall_out {
        I2_noErrCall_out { result: format!("result_{}", arg.value) }
    }
}

#[test]
fn i1_json_test_call_success() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let input_json = serde_json::to_string(&I1_testCall_in {}).unwrap();
    let rt = OutcomeServiceRt;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_I1(&method, &input_json, &MockI1, &rt, &ctx);
    match result {
        Outcome::Success(json_str) => {
            let decoded: I1_testCall_out = serde_json::from_str(&json_str).unwrap();
            assert_eq!(decoded.i00, 42);
        }
        Outcome::Failure(err) => panic!("Expected Success, got Failure({:?})", err.downcast_ref::<BaboonWiringError>()),
    }
}

#[test]
fn i1_json_test_call2_success() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall2".to_string() };
    let input_json = serde_json::to_string(&T7_Empty {}).unwrap();
    let rt = OutcomeServiceRt;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_I1(&method, &input_json, &MockI1, &rt, &ctx);
    assert!(matches!(result, Outcome::Success(_)));
}

#[test]
fn i1_json_domain_error() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let input_json = serde_json::to_string(&I1_testCall_in {}).unwrap();
    let rt = OutcomeServiceRt;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_I1(&method, &input_json, &FailingI1, &rt, &ctx);
    match result {
        Outcome::Failure(err) => {
            assert!(err.downcast_ref::<BaboonWiringError>().is_some(), "Expected BaboonWiringError::CallFailed");
        }
        Outcome::Success(_) => panic!("Expected Failure(CallFailed)"),
    }
}

#[test]
fn i1_json_unknown_method() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "nonexistent".to_string() };
    let rt = OutcomeServiceRt;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_I1(&method, "{}", &MockI1, &rt, &ctx);
    match result {
        Outcome::Failure(err) => {
            assert!(err.downcast_ref::<BaboonWiringError>().is_some(), "Expected BaboonWiringError");
        }
        Outcome::Success(_) => panic!("Expected Failure(NoMatchingMethod)"),
    }
}

#[test]
fn i1_json_bad_input() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let rt = OutcomeServiceRt;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_I1(&method, "not valid json!!", &MockI1, &rt, &ctx);
    match result {
        Outcome::Failure(err) => {
            assert!(err.downcast_ref::<BaboonWiringError>().is_some(), "Expected BaboonWiringError::DecoderFailed");
        }
        Outcome::Success(_) => panic!("Expected Failure(DecoderFailed)"),
    }
}

#[test]
fn i1_ueba_test_call_success() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let ctx = BaboonCodecContext::Default;
    let rt = OutcomeServiceRt;
    let mut writer = Vec::new();
    BaboonBinEncode::encode_ueba(&I1_testCall_in {}, &ctx, &mut writer).unwrap();
    let result = invoke_ueba_I1(&method, &writer, &MockI1, &rt, &ctx);
    match result {
        Outcome::Success(bytes) => {
            let mut cursor = std::io::Cursor::new(bytes);
            let decoded: I1_testCall_out = BaboonBinDecode::decode_ueba(&ctx, &mut cursor).unwrap();
            assert_eq!(decoded.i00, 42);
        }
        Outcome::Failure(err) => panic!("Expected Success, got Failure({:?})", err.downcast_ref::<BaboonWiringError>()),
    }
}

#[test]
fn i1_ueba_unknown_method() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "nonexistent".to_string() };
    let rt = OutcomeServiceRt;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_ueba_I1(&method, &[], &MockI1, &rt, &ctx);
    assert!(matches!(result, Outcome::Failure(_)));
}

#[test]
fn i2_json_no_err_call_success() {
    let method = BaboonMethodId { service_name: "I2".to_string(), method_name: "noErrCall".to_string() };
    let input_json = serde_json::to_string(&I2_noErrCall_in { value: 123 }).unwrap();
    let rt = OutcomeServiceRt;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_I2(&method, &input_json, &MockI2, &rt, &ctx);
    match result {
        Outcome::Success(json_str) => {
            let decoded: I2_noErrCall_out = serde_json::from_str(&json_str).unwrap();
            assert_eq!(decoded.result, "result_123");
        }
        Outcome::Failure(err) => panic!("Expected Success, got Failure({:?})", err.downcast_ref::<BaboonWiringError>()),
    }
}

#[test]
fn i2_ueba_no_err_call_success() {
    let method = BaboonMethodId { service_name: "I2".to_string(), method_name: "noErrCall".to_string() };
    let ctx = BaboonCodecContext::Default;
    let rt = OutcomeServiceRt;
    let mut writer = Vec::new();
    BaboonBinEncode::encode_ueba(&I2_noErrCall_in { value: 456 }, &ctx, &mut writer).unwrap();
    let result = invoke_ueba_I2(&method, &writer, &MockI2, &rt, &ctx);
    match result {
        Outcome::Success(bytes) => {
            let mut cursor = std::io::Cursor::new(bytes);
            let decoded: I2_noErrCall_out = BaboonBinDecode::decode_ueba(&ctx, &mut cursor).unwrap();
            assert_eq!(decoded.result, "result_456");
        }
        Outcome::Failure(err) => panic!("Expected Success, got Failure({:?})", err.downcast_ref::<BaboonWiringError>()),
    }
}
